### DISCUSSION TITLE
how to create a pipleline with Variable pixdim

### DISCUSSION THREAD

**Discussion by lg2578** [Project-MONAI/MONAI#4403] [Q&A]
I readed some examples about spacing transform.like
            Spacingd(
                keys=["image", "label"],
                pixdim=(1.0, 1.0, 1.0),
                mode=("bilinear", "nearest"),
            ),
I need to know a image's spacing first,and pixdim fixed in the transforms pipeline.
how to create a pipleline with Variable pixdim, because images put into transforms pipeline may have different spacing. 
I find there is affine matrix when loadimage,still I don't know how to do it.

**Comment by Nic-Ma** [comment]
Hi @ericspod @dongyang0122 ,

Could you please help share some best practices for this question?

Thanks in advance.

**Comment by dongyang0122** [comment]
@lg2578 the transform `Spacingd` converts the volumes to ones with the target spacing (fixed spacing, (1.0, 1.0, 1.0) in your example). It can take volumes with various spacing as input. not sure if this is what you asked.

**Reply by lg2578** [reply]
@dongyang0122  @Nic-Ma ,I've know a little,maybe, if there are  two images, one is 300*300,and spacing=(1.5,1.5),the other is 400*400,with spacing=(2,2), then set spacing to (1,1),will the two images output shape 200*200 ,it's right?
but when I tried it ,get some errors.
my code is here:    https://www.kaggle.com/lvgang623/monai-test
code:
```
trans=T.Compose([
    T.LoadImage(),
    T.AddChannel(),
    T.Spacing(pixdim=(1,1,3)),
   
])
path='../input/3d-solution-with-monai-produce-3d-data/train/case101/case101_day22/case101_day22_image.nii.gz'
img=trans(path)
print(img[1]['pixdim'])
img[0].shape
```
error:
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
/kaggle/input/monai-v081/monai/transforms/transform.py in apply_transform(transform, data, map_items, unpack_items, log_stats)
     87         if isinstance(data, (list, tuple)) and map_items:
---> 88             return [_apply_transform(transform, item, unpack_items) for item in data]
     89         return _apply_transform(transform, data, unpack_items)

/kaggle/input/monai-v081/monai/transforms/transform.py in <listcomp>(.0)
     87         if isinstance(data, (list, tuple)) and map_items:
---> 88             return [_apply_transform(transform, item, unpack_items) for item in data]
     89         return _apply_transform(transform, data, unpack_items)

/kaggle/input/monai-v081/monai/transforms/transform.py in _apply_transform(transform, parameters, unpack_parameters)
     52 
---> 53     return transform(parameters)
     54 

/kaggle/input/monai-v081/monai/transforms/utility/array.py in __call__(self, img)
    187         """
--> 188         return img[None]
    189 

KeyError: None

The above exception was the direct cause of the following exception:

RuntimeError                              Traceback (most recent call last)
/tmp/ipykernel_33/871096864.py in <module>
      6 ])
      7 path='../input/3d-solution-with-monai-produce-3d-data/train/case101/case101_day22/case101_day22_image.nii.gz'
----> 8 img=trans(path)
      9 print(img[1]['pixdim'])
     10 img[0].shape

/kaggle/input/monai-v081/monai/transforms/compose.py in __call__(self, input_)
    171     def __call__(self, input_):
    172         for _transform in self.transforms:
--> 173             input_ = apply_transform(_transform, input_, self.map_items, self.unpack_items, self.log_stats)
    174         return input_
    175 

/kaggle/input/monai-v081/monai/transforms/transform.py in apply_transform(transform, data, map_items, unpack_items, log_stats)
    111             else:
    112                 _log_stats(data=data)
--> 113         raise RuntimeError(f"applying transform {transform}") from e
    114 
    115 

RuntimeError: applying transform <monai.transforms.utility.array.AddChannel object at 0x7fe7c572a9d0>
